### PR TITLE
[Form][2.7] Allow choices with duplicated content

### PR DESCRIPTION
**Pull request by ewgRa**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no, but not sure |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15606 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by ewgRa** [comment]
Test fails seems not related.

ping @webmozart 

@symfony/deciders is it possible to have your point of view on this ?

File: src/Symfony/Component/Form/ChoiceList/ChoiceListInterface.php [hunk 1]
```diff
@@ -89,6 +89,27 @@ public function getStructuredValues();
     public function getOriginalKeys();
 
     /**
+     * Returns the raw choices.
+     *
+     * @return array The raw choices
+     */
+    public function getRawChoices();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/ChoiceListInterface.php]: adding a method to an interface is a BC break

Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/ChoiceListInterface.php]: maybe you could add a new interface instead, with only these methods

Comment (ewgRa) [inline src/Symfony/Component/Form/ChoiceList/ChoiceListInterface.php]: Ok, I will check.

**Comment by stof** [comment]
What is the use case for having duplicate choices ?
For duplicate labels, this can be resolved with the existing system by using ``choice_label`` to build labels for each choice instead of relying on the default behavior

**Comment by ewgRa** [comment]
@stof you can check #15606 for cases, there is some. Translations cases, grouped choices with same options in groups.

What is a choice_label is? Can you give a link?

Also it is a bug. To say people than it is "can be resolved by" - it must be documented as feature then. No?

**Comment by Tobion** [comment]
What @stof means and what I agree with is that duplicate choices (what you would get as selected value) makes no sense as you would not be able to distinguish each one anyway. Having the same label can happen and this is possible with the `choice_label` option.
The problem in #15606 is that the BC layer for the old choices structure (when `choices_as_values: false`) flips the choices which of course removes all choices which had the same label. So this BC layer must be fixed. So the current solution seems wrong.

**Comment by ewgRa** [comment]
@Tobion first of all thanks for review

> makes no sense as you would not be able to distinguish each one anyway

Sometimes it is not needed, sometimes business want to show same thing twice, just to increase chance that user select it, or do not miss. 

Also as check my case: https://github.com/symfony/symfony/issues/15606#issuecomment-152555285.

About choice_label, if you mean something like this:

```
    public function testCreateViewDuplicateArrayKeyChoiceListValuesWithChoiceLabel()
    {
        $list = new ArrayKeyChoiceList(
            array(
                'A' => 'a',
                'AA' => 'a'
            )
        );

        $view = $this->factory->createView(
            $list,
            null,
            function ($allChoices, $currentChoiceKey) {
                return 'a';
            }
        );

        $this->assertEquals(
            new ChoiceListView(
                array(
                    0 => new ChoiceView('A', 'A', 'a'),
                    1 => new ChoiceView('AA', 'AA', 'a')
                ),
                array()
            ),
            $view
        );
    }
```

It is not working.


> The problem in #15606 is that the BC layer for the old choices structure (when choices_as_values: false)
> So this BC layer must be fixed

Can you check tests from this PR? For example there is a test that create ArrayKeyChoiceList without second argument, and in this case useChoicesAsValues will be true. It is not only about BC layer as I understand.


Would be nice to hear @webmozart, as contributor of this code.

**Comment by webmozart** [comment]
Hi :) Thanks for working on this! I think this should be fixed in a less intrusive way. See #16685.

**Pull request closed without merging**